File: app/src/main/java/com/estebanlamas/planesradar/presentation/map/MapPresenter.java
package com.estebanlamas.planesradar.presentation.map;

import com.estebanlamas.planesradar.domain.GetAircraftsUseCase;
import com.estebanlamas.planesradar.domain.model.Aircraft;
import com.estebanlamas.planesradar.presentation.Presenter;

import java.util.List;
import java.util.Timer;
import java.util.TimerTask;


import javax.inject.Inject;

import rx.Subscription;

/**
 * Created by esteban on 24/9/17
 */

public class MapPresenter extends Presenter<MapView> {
    private static final int REFRESH_MILLISECONDS = 3000;
    private Timer refreshTimer = new Timer();
    private GetAircraftsUseCase getAircraftsUseCase;
    private Subscription getAircraftsSubscription;

    @Inject
    public MapPresenter(GetAircraftsUseCase getAircraftsUseCase) {
        this.getAircraftsUseCase = getAircraftsUseCase;
    }

    public void onResume() {
        initRadar();
    }

    private void initRadar() {
        refreshTimer.schedule(new TimerTask() {
            @Override
            public void run() {
                requestAircrafts();
            }
        }, 0, REFRESH_MILLISECONDS);
    }

    private void requestAircrafts() {
        getAircraftsSubscription = getAircraftsUseCase.execute()
                .subscribe(this::onAircraftReceived, this::showErrorView);
    }

    private void onAircraftReceived(List<Aircraft> aircraftList){
        view.updateAircrafts(aircraftList);
    }

    public void showErrorView(Throwable error) {

    }

    public void onPause() {
        getAircraftsSubscription.unsubscribe();
        refreshTimer.cancel();
    }
}

File: app/src/main/java/com/estebanlamas/planesradar/presentation/map/MapPresenter.kt
package com.estebanlamas.planesradar.presentation.map

import com.estebanlamas.planesradar.domain.GetAircraftsUseCase
import com.estebanlamas.planesradar.domain.model.AircraftsDetected
import com.estebanlamas.planesradar.presentation.Presenter
import rx.Subscription
import java.util.*
import javax.inject.Inject

/**
 * Created by esteban on 24/9/17
 */

class MapPresenter
@Inject constructor(private val getAircraftsUseCase: GetAircraftsUseCase) : Presenter<MapView>() {
    private val refreshTimer = Timer()
    private var getAircraftsSubscription: Subscription? = null

    fun onResume() {
        initRadar()
    }

    private fun initRadar() {
        refreshTimer.schedule(object : TimerTask() {
            override fun run() {
                requestAircrafts()
            }
        }, 0, REFRESH_MILLISECONDS)
    }

    private fun requestAircrafts() {
        getAircraftsSubscription = getAircraftsUseCase.execute()
                .subscribe({ this.onAircraftReceived(it) }, { this.showErrorView(it) })
    }

    private fun onAircraftReceived(aircraftsDetected: AircraftsDetected) {
        view?.updateAircrafts(aircraftsDetected.aircrafts)
    }

    fun showErrorView(error: Throwable) {
        view?.showError(error.message)
    }

    fun onPause() {
        getAircraftsSubscription?.unsubscribe()
        refreshTimer.cancel()
    }

    companion object {
        const val REFRESH_MILLISECONDS = 3000L
    }
}
